Support Refresh Tokens

rajadain opened the pull request:
## Overview

Refresh tokens are useful for CLIs, mobile devices, and other long-lived clients of the API. The idea is to generate a token tied to the client that allows the user to not have to specify their username and password every time.

The workflow for the user will be as follows:

 1. Generate a Refresh Token from Auth0
 2. Use the Refresh Token to get a JWT by POSTing the Refresh Token to `/api/tokens`

        {
          "refresh_token": "eyJ..."
        }

 3. Use the JWT in all their subsequent requests

Each Refresh Token has two pieces of metadata: an `id` and a `device_name`. The `id` is for uniquely identify the token, for the purpose of deletion. The `device_name` is a friendly name the user chooses when creating a Refresh Token, to help them remember which device it was for. Once generated, the actual Refresh Token is never saved anywhere except for on the device by the user. We only have access to it's metadata, not the token itself. Thus, if a user loses their Refresh Token, they must get a new one.

Users can get a list of active tokens by simply GETing `/api/tokens` with their JWT. To revoke a particular token, they can DELETE `/api/tokens/{id}`.

All the tokens and their authorization logic is handled by Auth0. We simply proxy the requests to them, since listing and deleting Refresh is done via the Management API which requires Client Secret authentication.

### Demo

#### Acquiring a Refresh Token

Ensure that _nothing_ is running on `http://localhost:9091`. Construct and visit the following URL:

```
https://raster-foundry.auth0.com/authorize?response_type=token&client_id={{ AUTH0_CLIENT_ID }}&connection=Username-Password-Authentication&redirect_uri=http://localhost:9091&device={{ YOUR_DEVICE_NAME }}&scope=openid%20offline_access
```

You should be redirected to `http://localhost:9091/#access_token=Vhj...&refresh_token=VSW...&id_token=eyJ...&token_type=Bearer` with the `refresh_token` in the query hash:

![image](https://cloud.githubusercontent.com/assets/1430060/19391230/ad67787c-91f9-11e6-94c1-b600c08532ca.png)


#### Getting a JWT with a Refresh Token

Acquiring a JWT requires only a valid Refresh Token, and no JWT need be specified in the `Authorization: Bearer eyJ...` header. The JWT received in `id_token` should be used in all future requests.

```http
> http :9000/api/tokens refresh_token=AgmvVAbrl38KJH15Fpx2yCkEBCt8LCDcDD5mbwWn6Ud1N
HTTP/1.1 200 OK
Content-Length: 414
Content-Type: application/json
Date: Fri, 14 Oct 2016 14:17:22 GMT
Server: akka-http/2.4.10

{
    "expires_in": 36000,
    "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDY0MiwiaWF0IjoxNDc2NDU0NjQyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9._UFhSJI7B7s4friZ8XjIqdMBu0lWp20n7x7shjILAZg",
    "token_type": "Bearer"
}

```

or, if you inexplicably prefer `curl`:

```
> curl -X POST -H "Content-Type: application/json" -d '{"refresh_token": "AgmvVAbrl38KJH15Fpx2yCkEBCt8LCDcDD5mbwWn6Ud1N"}' http://localhost:9000/api/tokens
{
  "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDk2MiwiaWF0IjoxNDc2NDU0OTYyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.nq7eDdCmAfxe_GhckoGO-bEEevwr0cjEJNYLDjNlyQQ",
  "expires_in": 36000,
  "token_type": "Bearer"
}
```

#### Listing Active Refresh Tokens

Here the user must be logged in, and use a valid JWT in the `Authorization: Bearer eyJ...` header.

```http
> http :9000/api/tokens "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDk2MiwiaWF0IjoxNDc2NDU0OTYyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.nq7eDdCmAfxe_GhckoGO-bEEevwr0cjEJNYLDjNlyQQ"
HTTP/1.1 200 OK
Content-Length: 80
Content-Type: application/json
Date: Fri, 14 Oct 2016 14:24:39 GMT
Server: akka-http/2.4.10

[
    {
        "device_name": "ttuhinanshu-macbook-pro",
        "id": "dcr_iGZTCZOUpfAMvof4"
    }
]
```

```
> curl -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDk2MiwiaWF0IjoxNDc2NDU0OTYyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.nq7eDdCmAfxe_GhckoGO-bEEevwr0cjEJNYLDjNlyQQ" http://localhost:9000/api/tokens
[{
  "id": "dcr_iGZTCZOUpfAMvof4",
  "device_name": "ttuhinanshu-macbook-pro"
}]
```

#### Revoking Refresh Token

```http
> http DELETE :9000/api/tokens/dcr_iGZTCZOUpfAMvof4 "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDk2MiwiaWF0IjoxNDc2NDU0OTYyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.nq7eDdCmAfxe_GhckoGO-bEEevwr0cjEJNYLDjNlyQQ"
HTTP/1.1 204 No Content
Content-Type: text/plain; charset=UTF-8
Date: Fri, 14 Oct 2016 14:27:35 GMT
Server: akka-http/2.4.10

```

```
> curl -X DELETE -H  "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5MDk2MiwiaWF0IjoxNDc2NDU0OTYyLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.nq7eDdCmAfxe_GhckoGO-bEEevwr0cjEJNYLDjNlyQQ" http://localhost:9000/api/tokens/dcr_iGZTCZOUpfAMvof4

```

### Notes

The central question here is whether we pipe through the Auth0 responses directly, at a lower processing cost but potentially exposing more information than we mean to, or we intercept and reformat things, which comes at a cost of not just some performance but also code complexity, for the sake of clean interfaces?

With interception, a response looks like this:

```http
HTTP/1.1 200 OK
Content-Length: 414
Content-Type: application/json
Date: Fri, 14 Oct 2016 16:02:07 GMT
Server: akka-http/2.4.10

{
    "expires_in": 36000,
    "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5NjkyNywiaWF0IjoxNDc2NDYwOTI3LCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.RYxFdjZKJ-ZLWn0l094Spzw73lMGOHTRrJWt4WgXxkg",
    "token_type": "Bearer"
}
```

without it, it looks like this:

```http
HTTP/1.1 200 OK
Cache-Control: no-cache
Connection: keep-alive
Content-Length: 401
Content-Type: application/json; charset=utf-8
Date: Fri, 14 Oct 2016 16:03:03 GMT
Keep-Alive: timeout=100
Set-Cookie: auth0=wZl-JBTGEnzAW08NEiNRuQ.9OjacqFY7_bjWY3yy7PwFop0wMSYCvpFwW0BWsyhaHCzUKLiTxXvNWzG96y5DR3x.1476460983022.604800000.IrHg-qWfjqbSqgXQGoonE2TKV8GpOKF38m6qD977T0c; path=/; expires=Mon, 17 Oct 2016 16:03:03 GMT; secure; httponly
Strict-Transport-Security: max-age=15724800
X-RateLimit-Limit: 10
X-RateLimit-Remaining: 8
X-RateLimit-Reset: 1476461048
X-Robots-Tag: noindex, nofollow, nosnippet, noarchive

{
    "expires_in": 36000,
    "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1N2Y3YjMwMGI3ZWZkZTY5NmI0MjI0NTMiLCJhdWQiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSIsImV4cCI6MTQ3NjQ5Njk4MywiaWF0IjoxNDc2NDYwOTgzLCJhenAiOiJaYUN4eWlraURMRkxLQk9mUEI1UjMwdG5jOHIwNm54VSJ9.DR4p3D8caegYSQ8EUl2PEinldjjn4S_ZMpmORu-Txmg",
    "token_type": "Bearer"
}
```

## Testing Instructions

 * Check out this branch
 * Get new credentials from S3:

        $ aws s3 --profile=raster-foundry cp s3://config-development-raster-foundry/.env .env

 * Spin up the server and the front-end, and ensure that you can login with your personal Auth0 account
 * Shut down the front-end
 * Go to https://raster-foundry.auth0.com/authorize?response_type=token&client_id=AUTH0_CLIENT_ID&connection=Username-Password-Authentication&redirect_uri=http://localhost:9091&device=YOUR_DEVICE_NAME&scope=openid%20offline_access after replacing `AUTH0_CLIENT_ID` and `YOUR_DEVICE_NAME` in the URL appropriately.
 * You should be redirected to `http://localhost:9091/#/` with `access_token`, `refresh_token`, `id_token`, and `token_type` in the hash. Copy the `refresh_token` to a file.
 * Ensure the back-end is up
 * Test getting JWT with your `refresh_token`

        $ http :9000/api/tokens refresh_token=YOUR_REFRESH_TOKEN

 * Test listing all your Refresh Tokens with the JWT from the previous call

        $ http :9000/api/tokens "Authorization: Bearer eyJ..."

 * Test deleting the Refresh Token returned by the previous call

        $ http :9000/api/tokens/dcr_... "Authorization: Bearer eyJ..."

Connects #523

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+            val req = Http().singleRequest(HttpRequest(GET, uri.withQuery(params), headers))
+
+            onSuccess(req) {
+              case HttpResponse(OK, _, entity, _) => complete(Unmarshal(entity).to[List[DeviceCredential]])
```
The interception here is not completely useless, even though its ultimate utility is questionable. The response we get from Auth0 is of the format:

```json
[
    {
        "device_name": "ttuhinanshu-macbook-pro",
        "id": "dcr_EEUkyRHihMspypXG",
        "type": "refresh_token",
        "user_id": "auth0|57f..."
    }
]
```

Of these, `id` and `device_name` are legitimately valuable. The `type` will always be `refresh_token`, and the `user_id` will always be the user's id, thus there is no use sending these to the client. However, while those bits are useless, they are also harmless, and we may be doing extra work to remove them. It is worth considering if this interception is justifiable, and whether we should just pipe the request through.

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+              ))
+
+              onSuccess(req) {
+                case HttpResponse(OK, _, entity, _) => complete(Unmarshal(entity).to[AuthorizedToken])
```
Similar to above but less justifiable: this unmarshals the entity into exactly itself. The whole purpose of this intercept is to hide the headers, making Auth0 invisible. If we were to let it pass through, only the headers would be different.

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+          delete {
+            // TODO Check that the deviceId belongs to user before deleting it!
+            // Otherwise any authenticated user will be able to delete any
+            // device key, whether or not they own it.
```
I think this is really important, but couldn't think of an elegant way to solve this without creating a blocking HTTP request that first checks for the user's tokens. Also couldn't really make that request because I got really confused by the `Future` types. But I don't think we should merge this without a solution for this.

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+
+import com.azavea.rf.auth.Authentication
+import com.azavea.rf.utils.{Config, UserErrorHandler}
+
```
That's an impressive amount of magic being imported right here

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+            ))
+
+            onSuccess(req) {
+              case HttpResponse(OK, _, entity, _) => complete(Unmarshal(entity).to[AuthorizedToken])
```
Similar to above but less justifiable: this unmarshals the entity into exactly itself. The whole purpose of this intercept is to hide the headers, making Auth0 invisible. If we were to let it pass through, only the headers would be different.

rajadain commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+          authenticate { user =>
+            // TODO Check that the deviceId belongs to user before deleting it!
+            // Otherwise any authenticated user will be able to delete any
+            // device key, whether or not they own it.
```
I think this is really important, but couldn't think of an elegant way to solve this without creating a blocking HTTP request that first checks for the user's tokens. Also couldn't really make that request because I got really confused by the `Future` types. But I don't think we should merge this without a solution for this.

notthatbreezy replied:
We can check this out -- if everything is wrapped in a `Future` we should be fine, I have some ideas here.

notthatbreezy commented on app-server/app/src/main/resources/application.conf:
```diff
@@ -53,4 +53,7 @@ auth0 {
   secret = ${?AUTH0_CLIENT_SECRET}
+  # Auth0 Bearer with proper permissions for Management API
+  bearer = ""
+  bearer = ${?AUTH0_MANAGEMENT_BEARER}
```
If we need permissions for other things does it make sense to update this token or create a different one?

rajadain replied:
Another potential point of failure is that if we update this token to not have all the permissions the code needs, there may be runtime failures. I wonder if we can make a test of some sort that decodes the token (it is simple JWT) and ensures that the right permissions are enabled:

```json
{
  "aud": "UlV...",
  "scopes": {
    "device_credentials": {
      "actions": [
        "read",
        "create",
        "delete",
        "update"
      ]
    }
  },
  "iat": 147000000,
  "jti": "dd1..."
}
```

notthatbreezy commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+  )
+
+  def tokenRoutes: Route = pathPrefix("api" / "tokens") {
+    handleExceptions(userExceptionHandler) {
```
I think it makes sense to split up the token routes and stack them such that we do something like [`thumbnailRoutes`](https://github.com/azavea/raster-foundry/blob/develop/app-server/app/src/main/scala/thumbnail/Routes.scala#L28) to make this a little more clearer.

rajadain replied:
I saw some routes that were defined inline, and some that were segregated as such. Wasn't sure which pattern to use. Will update to use this instead.

notthatbreezy commented on app-server/app/src/main/scala/token/Routes.scala:
```diff
@@ -0,0 +1,102 @@
+      pathEndOrSingleSlash {
+        get {
+          authenticate { user =>
+            val params = Query(
```
Does it make sense to split some of the logic in these routes into a separate service and keep _only_ routing here in the router?

rajadain replied:
I can do that. It would certainly make the routing cleaner. I thought that the `...Service.scala` pattern was going away with #516, but if it makes sense here I'll add that.

notthatbreezy replied:
It only goes away in #516 insofar as the database logic has been moved out -- I think we'll see use cases for it in the future though.

rajadain commented:
I covered everything discussed today, but the changes are incompatible with the latest refactor. I've rebased on that, but I seem to have lost the execution context:

```
[error] /opt/raster-foundry/app-backend/app/src/main/scala/token/Routes.scala:23: Cannot find an implicit ExecutionContext. You might pass
[error] an (implicit ec: ExecutionContext) parameter to your method
[error] or import scala.concurrent.ExecutionContext.Implicits.global.
[error]         onSuccess(TokenService.listRefreshTokens(user)) { tokens =>
[error]                                                 ^
```

I'm not sure where to import it from. Will see if this can be resolved tomorrow.

kshepard commented:
We're using the global execution context now. You can get it with: 
`import scala.concurrent.ExecutionContext.Implicits.global`.

rajadain commented:
Rebased on top of latest `develop`. Added an `Auth0Exception` which takes the error code and text returned from the Auth0 API. This can be tested by commenting out `AUTH0_MANAGEMENT_BEARER` in `.env`. The end user sees the error code in their response, and get a code of 500 themselves (since this is between our backend and Auth0, a true server error):

```http
> http :9000/api/tokens "Authorization: Bearer eyJ..."
HTTP/1.1 500 Authentication Service Error
Content-Length: 53
Content-Type: text/plain; charset=UTF-8
Date: Tue, 18 Oct 2016 15:08:23 GMT
Server: akka-http/2.4.11

Authentication service returned error 400 Bad Request
```

The full error is logged to the backend console / output:

```
[info] 15:08:23.762 [rf-system-akka.actor.default-dispatcher-7] ERROR com.azavea.rf.Main$ - com.azavea.rf.token.Auth0Exception: HttpEntity.Strict(application/json,{"statusCode":400,"error":"Bad Request","message":"Bad HTTP authentication header format","errorCode":"Bearer"})
[info] 	at com.azavea.rf.token.TokenService$$anonfun$listRefreshTokens$1.apply(Token.scala:47)
[info] 	at com.azavea.rf.token.TokenService$$anonfun$listRefreshTokens$1.apply(Token.scala:43)
[info] 	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
[info] 	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
[info] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
[info] 	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
[info] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[info] 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I'm using the [`PrintWriter` technique](http://alvinalexander.com/scala/how-convert-stack-trace-exception-string-print-logger-logging-log4j-slf4j) for outputting the stack trace, since doing `e.getStackTrace.toString` as done [elsewhere](https://github.com/azavea/raster-foundry/blob/13ac14022fc9d9ce07c096968d2be57d440ef20d/app-backend/app/src/main/scala/utils/UserErrorHandler.scala#L11) results in the uninformative:

```
[info] 15:07:16.346 [rf-system-akka.actor.default-dispatcher-9] ERROR com.azavea.rf.Main$ - [Ljava.lang.StackTraceElement;@5a374541
```

I added `RfStackTrace` to `com.azavea.rf.utils` so it can be used throughout the codebase. If there is a more Scala-idiomatic way to do it, we will just have to change one spot.

"Auth0" itself is never mentioned in any communication with the end user.

rajadain commented:
Also changed from

```scala
authenticate { user =>
  get {
    onSuccess(TokenService.listRefreshTokens(user)) { tokens =>
      complete(tokens)
    }
  }
}
```

to

```scala
get {
  authenticate { user =>
    onSuccess(TokenService.listRefreshTokens(user)) { tokens =>
      complete(tokens)
    }
  }
}
```

In the first case, since we were authenticating before allowing `GET`, the wrong credentials would return a `405 Method Not Allowed`. Now, since we declare `GET` first and check the credentials second, hitting the right end point with the wrong credentials correctly returns `401 Not Authorized`. Same for the `DELETE` endpoint.
